src/util/chunked_content_decoder.js

```diff
@@ -41,57 +64,199 @@ class ChunkedContentDecoder extends stream.Transform {
     }
 
     _flush(callback) {
-        if (this.state !== STATE_CONTENT_END) return this.error_state();
+        if (this.state !== STATE_CONTENT_END) return this.error_state(undefined, 0);
         return callback();
     }
 
+    /**
+     * Parse the buffer and update the state machine.
+     * The buffer is parsed in a loop to handle multiple chunks in the same buffer,
+     * and to handle the case where the buffer ends in the middle of a chunk.
+     * The state machine is updated according to the current state and the buffer content.
+     * The state machine is updated by the following rules:
+     *   1. STATE_READ_CHUNK_HEADER - read the chunk header until CR and parse it.
+     *   2. STATE_WAIT_NL_HEADER - wait for NL after the chunk header.
+     *   3. STATE_SEND_DATA - send chunk data to the stream until chunk size bytes sent.
+     *   4. STATE_WAIT_CR_DATA - wait for CR after the chunk data.
+     *   5. STATE_WAIT_NL_DATA - wait for NL after the chunk data.
+     *   6. STATE_READ_TRAILER - read optional trailer until CR and save it.
+     *   7. STATE_WAIT_NL_TRAILER - wait for NL after non empty trailer.
+     *   8. STATE_WAIT_NL_END - wait for NL after the last empty trailer.
+     *   9. STATE_CONTENT_END - the stream is done.
+     *  10. STATE_ERROR - an error occurred.
+     * @param {Buffer} buf
+     * @returns {boolean} false on error state
+     */
     parse(buf) {
         for (let index = 0; index < buf.length; ++index) {
+
+            //---------------//
+            // header states //
+            //---------------//
+
             if (this.state === STATE_READ_CHUNK_HEADER) {
-                for (; index < buf.length; ++index) {
-                    if (buf[index] === CR_CODE) {
-                        const header_items = this.chunk_header_str.split(';');
-                        this.chunk_size = parseInt(header_items[0], 16);
-                        if (!(this.chunk_size >= 0)) return this.error_state();
-                        this.last_chunk = this.chunk_size === 0;
-                        const header1 = header_items[1].split('=');
-                        this.chunk_signature = header1[0] === 'chunk-signature' ? header1[1] : '';
-                        this.chunk_header_str = '';
-                        this.state = STATE_WAIT_NL_HEADER;
-                        break;
-                    } else {
-                        this.chunk_header_str += String.fromCharCode(buf[index]);
-                    }
+                const { str, next, finished } = this.read_string_until_cr(buf, index);
+                index = next;
+                this.chunk_header += str;
+                if (this.chunk_header.length > MAX_CHUNK_HEADER_SIZE) {
+                    return this.error_state(buf, index);
```

Comment:
maybe we should also add an explantion about why we failing, like here, for example:
`chunk header size bigger than max allowed` and the same for the rest

Reply:
these are just safety limits which aim to prevent reading data that does not follow the encoding. we can add a message to the error_state(), but since this is deep inside the content encoding it will still be hard trace no matter what we say, and the root cause will likely just be that the incoming content is malformed, or our code is bad, so nothing that the user can understand.

Reply:
I added the reason to be printed, hoping it would help developers.